### PR TITLE
feat: OLM operator installer

### PR DESCRIPTION
**Pull request by AndrienkoAleksandr**
### What does this PR do?
Add ability to install Che using OLM installer. In the current state we install only stable channel. Also provided flags to use custom catalog source in the same namespace with Che installation.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/15279

### CODE REVIEW

File: src/tasks/installers/olm.ts [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+          ctx.defaultCatalogSourceNamespace = flags.platform === 'crc' ? defaultOpenshiftMarketPlaceNamespace : defaultOLMKubernetesNamespace
+          // preview package name
+          ctx.packageName = this.packageNamePrefix + (ctx.isOpenShift ? 'openshift' : 'kubernetes')
+          // catalog source name for stable Che version
+          ctx.catalogSourceNameStable = isKubernetesPlatformFamily(flags.platform) ? 'operatorhubio-catalog' : 'community-operators'
+          task.title = `${task.title}...OK`
+        }
+      },
+      {
+        title: 'Create custom catalog source for "nightly" channel...OK',
```
Comment (tolusha) [inline src/tasks/installers/olm.ts:64]: add `...OK` when task is completed

File: src/tasks/installers/olm.ts [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+  }
+
+  deleteTasks(flags: any): ReadonlyArray<Listr.ListrTask> {
+    const kube = new KubeHelper(flags)
+    return [
+      {
+        title: "Check if OLM is pre-installed on the platform",
+        task: async  (ctx: any, task: any) => {
+          ctx.isPreInstalledOLM = await kube.isPreInstalledOLM() ? true : false
+        }
```
Comment (tolusha) [inline src/tasks/installers/olm.ts:186]: just missed
`task.title = `${task.title}...OK``

File: src/tasks/installers/olm.ts [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+      },
+      {
+        title: 'Delete(OLM) Eclipse Che cluster service versions',
+        enabled: (ctx) => ctx.isPreInstalledOLM,
+        task: async (ctx: any, task: any) => {
+          const csvs = await kube.getClusterServiceVersions(flags.chenamespace)
+          const csvsToDelete = csvs.items.filter((csv) => csv.metadata.name.startsWith("eclipse-che"))
+          csvsToDelete.forEach((csv) => kube.deleteClusterServiceVersion(flags.chenamespace, csv.metadata.name))
```
Comment (tolusha) [inline src/tasks/installers/olm.ts:194]: just missed
`task.title = `${task.title}...OK``

**Comment by tolusha** [comment]
```
 Check if OLM is pre-installed on the platform
      → OLM isn't installed on your platfrom. If your platform hasn't got embedded OML, you need install it manually.
```
Why don't install it automatically? Are there any pitfalls?

**Comment by AndrienkoAleksandr** [comment]
> @tolusha  Why don't install it automatically? Are there any pitfalls?

@tolusha [I see in the docs of the OLM](https://github.com/operator-framework/operator-sdk/blob/master/doc/user/olm-catalog/olm-cli.md), new operator-sdk has command: 
`operator-sdk olm install` ... Maybe we should propose for user download operator-sdk binary and launch this command instead of clone vehicle? Also we have manual installation link https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/install/install.md

File: src/tasks/installers/olm.ts [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+      checkPreCreatedTls(flags, kube),
+      checkTlsSertificate(flags),
+      {
+        title: 'Create operator group',
+        task: async (task: any) => {
+          if (await kube.operatorGroupExists(this.operatorGroupName, flags.chenamespace)) {
+            task.title = `${task.title}...It already exists.`
+          } else {
+            await kube.createOperatorGroup(this.operatorGroupName, flags.chenamespace)
+            task.title = `${task.title}...OK`
```
Comment (mmorhun) [inline src/tasks/installers/olm.ts:46]: I think we usually use `done`. Please check chectl output.

Comment (AndrienkoAleksandr) [reply src/tasks/installers/olm.ts]: Looks like this... But for delete tasks we are using `...OK` for some unknown for me reasons..

**Comment by tolusha** [comment]
minikube 1.5.2
Installed olm using the following commands
```bash
kubectl apply -f https://raw.githubusercontent.com/operator-framework/operator-lifecycle-manager/master/deploy/upstream/quickstart/crds.yaml
 kubectl apply -f https://raw.githubusercontent.com/operator-framework/operator-lifecycle-manager/master/deploy/upstream/quickstart/olm.yaml
``bash

then

```bash
./run  server:start --platform minikube --installer olm 

 ›   Warning: Self-signed certificate is used, so "--self-signed-cert" option is required. Added automatically.
  ✖ Create operator source
    → [object Object]
    Create catalog source
  ❯ 🏃‍  Running Olm installaion Eclipse Che
    ✔ Check if OLM is pre-installed on the platform
    ✔ Copying operator resources...done.
    ✔ Create Namespace (che)...It already exists.
    ✔ Checking for pre-created TLS secret... "che-tls" secret found
    ✔ Checking certificate... self-signed
    ✔ Create operator group
    ✔ Configure context information...done.
    ✖ Create custom catalog source for "nightly" channel
      → [object Object]
      Create operator subscription
      Wait while subscription is ready
```

**Comment by tolusha** [comment]
Not all version of operator-sdk has `olm` command

```bash
$ operator-sdk version
operator-sdk version: v0.10.0, commit: ff80b17737a6a0aade663e4827e8af3ab5a21170
$ operator-sdk olm install
operator-sdk olm install
Error: unknown command "olm" for "operator-sdk"
```

**Comment by flacatus** [comment]
[test]

**Comment by AndrienkoAleksandr** [comment]
@tolusha Yes I see what you meant, looks like OperatorSource was removed from the latest olm... I will take a look to make it working... But any way creation OperatorSource it was workaround because in the operatorHub catalog for Che we don't have nightly channel...

**Comment by AndrienkoAleksandr** [comment]
[test]

**Comment by AndrienkoAleksandr** [comment]
[test]

**Comment by AndrienkoAleksandr** [comment]
[test]

**Comment by AndrienkoAleksandr** [comment]
[test]

**Comment by AndrienkoAleksandr** [comment]
[test]

**Pull request merged**